xla/stream_executor/command_buffer.h

```diff
@@ -172,177 +127,126 @@ class CommandBuffer {
-                             DeviceMemory<bool> pred,
-                             ExecutionScopeBuilder cond_builder,
-                             Builder body_builder) = 0;
-
-  // Adds a conditional While operation to default execution scope.
-  absl::Status While(DeviceMemory<bool> pred,
-                     ExecutionScopeBuilder cond_builder, Builder body_builder) {
-    return While(kDefaultExecutionScope, pred, cond_builder, body_builder);
-  }
+  // Create a new conditional handle in the underlying
+  // graph.
+  virtual absl::StatusOr<GraphConditionalHandle> CreateConditionalHandle() = 0;
+
+  // Adds a new conditional node to the graph and creates a
+  // corresponding nested command buffer.
+  virtual absl::StatusOr<ConditionalNodeResult> CreateConditionalNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle conditional,
+      ConditionType type) = 0;
+
+  virtual absl::StatusOr<GraphNodeHandle> CreateSetIfElseConditionKernelNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle then_condition,
+      GraphConditionalHandle else_condition, DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::Status UpdateSetIfElseConditionKernelNode(
+      GraphNodeHandle node, GraphConditionalHandle then_condition,
+      GraphConditionalHandle else_condition, DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::StatusOr<GraphNodeHandle> CreateSetIfConditionKernelNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle then_condition,
+      DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::Status UpdateSetIfConditionKernelNode(
+      GraphNodeHandle node, GraphConditionalHandle then_condition,
+      DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::StatusOr<GraphNodeHandle> CreateSetForConditionKernelNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle condition,
+      DeviceMemory<int32_t> loop_counter, int32_t iterations) = 0;
+
+  virtual absl::Status UpdateSetForConditionKernelNode(
+      GraphNodeHandle node, GraphConditionalHandle condition,
+      DeviceMemory<int32_t> loop_counter, int32_t iterations) = 0;
+
+  virtual absl::StatusOr<GraphNodeHandle> CreateSetWhileConditionKernelNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle condition,
+      DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::Status UpdateSetWhileConditionKernelNode(
+      GraphNodeHandle node, GraphConditionalHandle condition,
+      DeviceMemory<bool> predicate) = 0;
+
+  virtual absl::StatusOr<GraphNodeHandle> CreateSetCaseConditionKernelNode(
+      GraphNodeHandles dependencies, GraphConditionalHandle handle0,
+      GraphConditionalHandle handle1, GraphConditionalHandle handle2,
+      GraphConditionalHandle handle3, GraphConditionalHandle handle4,
+      GraphConditionalHandle handle5, GraphConditionalHandle handle6,
+      GraphConditionalHandle handle7, DeviceMemory<uint8_t> index,
+      bool index_is_bool, int32_t batch_offset, int32_t num_branches,
```

Comment:
can you infer index_is_bool from index device memory size?

Reply:
will create a separate PR for this.